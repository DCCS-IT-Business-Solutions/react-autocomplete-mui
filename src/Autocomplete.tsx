import * as React from "react";

import {
  TextField,
  List,
  ListItem,
  Popper,
  Paper,
  Fade,
  CircularProgress,
  InputAdornment
} from "@material-ui/core";
import Typography from "@material-ui/core/Typography";
import { TextFieldProps } from "@material-ui/core/TextField";
import { HighlightQuery } from "@dccs/utils";
import { useDebounce } from "use-debounce";

export interface IAutocompleteBaseProps {
  onOptionSelected?: (value: any) => void;
  onLoadOptions: (query: string) => Promise<any[]> | any[];
  renderOption?: (element: any, query: string) => React.ReactNode;
  renderNoOptionsFound?: () => React.ReactNode;
  maxShownOptions?: number;
  maxHeightOptionsList?: number;
  textProp?: (option: any) => string;
  valueProp?: (option: any) => string | number | any;
}

export type IAutocompleteProps = IAutocompleteBaseProps & TextFieldProps;

export function Autocomplete(props: IAutocompleteProps) {
  const newProps = { ...Autocomplete.defaultProps, ...props };
  const {
    onOptionSelected,
    onLoadOptions,
    renderOption,
    renderNoOptionsFound,
    maxShownOptions,
    maxHeightOptionsList,
    textProp,
    valueProp,
    value,
    ...others
  } = newProps;

  const [isFocused, setIsFocused] = React.useState<boolean>(false);
  const [loading, setLoading] = React.useState<boolean>(false);
  const [suggestions, setSuggestions] = React.useState<any[]>();
  const [textFieldValue, setTextFieldValue] = React.useState(
    textProp(value || "")
  );
  const [highlightedOption, setHighlightedOption] = React.useState<number>(0);
  const [debouncedQuery] = useDebounce<string>(textFieldValue, 500);

  const textFieldRef = React.useRef<HTMLDivElement>(null);

  function handleKeyDown(e: React.KeyboardEvent<HTMLDivElement>) {
    switch (e.keyCode) {
      case 40: // down arrow
        e.stopPropagation();
        if (
          (suggestions != null &&
            suggestions.length <= highlightedOption + 1) ||
          (maxShownOptions && maxShownOptions <= highlightedOption + 1)
        ) {
          setHighlightedOption(0);
        } else {
          setHighlightedOption(highlightedOption + 1);
        }
        break;
      case 38: // up arrow
        e.stopPropagation();
        if (highlightedOption === 0) {
          if (maxShownOptions) {
            setHighlightedOption(maxShownOptions - 1);
          } else if (suggestions) {
            setHighlightedOption(suggestions.length - 1);
          }
        } else {
          setHighlightedOption(highlightedOption - 1);
        }
        break;
      case 13: // enter
      case 9: // tab
        e.stopPropagation();
        e.preventDefault();
        if (suggestions) {
          const element = suggestions[highlightedOption];
          onOptionSelected(valueProp(element));
        }
        if (textFieldRef.current != null) {
          textFieldRef.current.blur();
        }
        setIsFocused(false);
        break;
      case 27: // esc
        e.stopPropagation();
        if (textFieldRef.current != null) {
          textFieldRef.current.blur();
        }
        setIsFocused(false);
        break;
    }
  }

  const allSuggestions = React.useMemo(() => {
    return onLoadOptions("");
  }, []);

  React.useEffect(() => {
    if (isFocused) {
      setLoading(true);
    }
  }, [textFieldValue]);

  async function loadOptions(query: string) {
    const result = await onLoadOptions(query || "");
    setSuggestions(result);
    setLoading(false);
  }

  async function setValueAsTextFieldValue() {
    setLoading(true);
    if (allSuggestions) {
      const result = await allSuggestions;
      const selectedSuggestions = result.find(o => valueProp(o) === value);
      if (selectedSuggestions) {
        setTextFieldValue(textProp(selectedSuggestions));
      }
    }
    setLoading(false);
  }

  React.useEffect(() => {
    if (value) {
      setValueAsTextFieldValue();
    }
  }, []);

  React.useEffect(() => {
    loadOptions(debouncedQuery);
  }, [debouncedQuery]);

  React.useEffect(() => {
    if (value) {
      if (valueProp) {
        setValueAsTextFieldValue();
      } else {
        setTextFieldValue(textProp(value));
      }
    } else {
      setTextFieldValue("");
    }
    setIsFocused(false);
  }, [value]);

  React.useEffect(() => {
    if (!isFocused) {
      if (value) {
        if (valueProp) {
          setValueAsTextFieldValue();
        } else {
          setTextFieldValue(textProp(value));
        }
      }
    }
  }, [isFocused]);

  return (
    <div onKeyDown={isFocused ? handleKeyDown : undefined}>
      <TextField
        ref={textFieldRef}
        InputProps={{
          endAdornment: loading ? (
            <InputAdornment position="end">
              <CircularProgress size={18} />
            </InputAdornment>
          ) : (
            undefined
          )
        }}
        style={{ minWidth: "500px" }}
        {...others}
        onFocus={e => {
          setIsFocused(true);
          onLoadOptions(textFieldValue);
        }}
        onBlur={(e: any) => {
          if (suggestions) {
            onOptionSelected(valueProp(suggestions[highlightedOption]));
          }

          setIsFocused(false);
        }}
        onChange={e => {
          setIsFocused(true);
          setTextFieldValue(e.target.value);
          onLoadOptions(e.target.value);
        }}
        value={textFieldValue || ""}
      />
      {isFocused && (
        <Popper
          open={isFocused}
          anchorEl={textFieldRef.current}
          placement="bottom-start"
          style={{ zIndex: 1000000 }}
          transition
        >
          {({ TransitionProps }) => (
            <Fade {...TransitionProps} timeout={10}>
              <OptionList
                loading={loading}
                suggestions={suggestions}
                highlightedOption={highlightedOption}
                setHighlightedOption={setHighlightedOption}
                onOptionSelected={onOptionSelected}
                renderOption={element =>
                  renderOption ? (
                    renderOption(element, textFieldValue)
                  ) : (
                    <Typography>
                      {HighlightQuery(textProp(element), textFieldValue)}
                    </Typography>
                  )
                }
                renderNoOptionsFound={renderNoOptionsFound}
                maxOptions={maxShownOptions}
                maxHeight={maxHeightOptionsList}
                valueProp={valueProp}
              />
            </Fade>
          )}
        </Popper>
      )}
    </div>
  );
}

Autocomplete.defaultProps = {
  onOptionSelected: (value: any) => {
    window.alert(value);
  },
  textProp: (option: any) => (option.value ? option.value : option.toString()),
  valueProp: (option: any) => (option.id ? option.id : option),
  maxShownOptions: 10,
  maxHeightOptionsList: 500,
  renderNoOptionsFound: () => <Typography>No matches found!</Typography>
};

interface IOptionListProps {
  loading: boolean;
  suggestions?: any[];
  maxOptions: number | undefined;

  onOptionSelected: (e: any) => void;

  highlightedOption: number;
  setHighlightedOption: (value: number) => any;
  renderOption: (element: any) => React.ReactNode;
  renderNoOptionsFound: () => React.ReactNode;
  maxHeight: number;
  valueProp: (option: any) => string | number | any;
}

function OptionList(props: IOptionListProps) {
  React.useEffect(() => {
    const selectedLi = document.getElementById(
      "react-autocomplete-mui-selected-option"
    );
    if (selectedLi) {
      setTimeout(
        () =>
          selectedLi.scrollIntoView({ block: "center", behavior: "smooth" }),
        5
      );
    }
  }, [props.highlightedOption]);

  return (
    <Paper elevation={6}>
      {props.suggestions && props.suggestions.length > 0 ? (
        <List
          style={{
            minWidth: 250,
            maxHeight: props.maxHeight || 500,
            overflow: "auto"
          }}
        >
          {props.suggestions
            .slice(0, props.maxOptions)
            .map((element: any, index: number) => (
              <OptionItem
                key={index}
                index={index}
                element={element}
                onOptionSelected={props.onOptionSelected}
                renderOption={props.renderOption}
                highlightedOption={props.highlightedOption}
                setHighlightedOption={props.setHighlightedOption}
                valueProp={props.valueProp}
              />
            ))}
        </List>
      ) : (
        !props.loading && (
          <ListItem key={"noMatch"} button>
            {props.renderNoOptionsFound()}
          </ListItem>
        )
      )}
    </Paper>
  );
}

interface IOptionItemProps {
  index: number;
  element: any;
  onOptionSelected: (value: any) => void;
  highlightedOption: number;
  setHighlightedOption: (value: number) => void;
  renderOption: (element: any) => React.ReactNode;
  valueProp: (option: any) => string | number | any;
}

function OptionItem(props: IOptionItemProps) {
  return React.useMemo(() => {
    return (
      <ListItem
        selected={props.index === props.highlightedOption}
        id={
          props.index === props.highlightedOption
            ? "react-autocomplete-mui-selected-option"
            : undefined
        }
        key={props.index}
        button
        onMouseDown={() => {
          props.setHighlightedOption(props.index);
        }}
        onClick={(e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
          e.stopPropagation();
          props.onOptionSelected(props.valueProp(props.element));
        }}
      >
        {props.renderOption(props.element)}
      </ListItem>
    );
  }, [props.highlightedOption, props.renderOption]);
}
