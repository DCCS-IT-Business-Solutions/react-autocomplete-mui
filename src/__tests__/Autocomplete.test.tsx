import * as React from "react";
import { shallow, mount, render } from "enzyme";
import { Autocomplete } from "../Autocomplete";
import { HighlightQuery } from "../HighlightQuery";

it("should render", () => {
  const sut = render(
    <Autocomplete value="" onLoadOptions={(query: string) => []} />
  );
  expect(sut).toMatchSnapshot();
});

// it("should call onRowClick if row is clicked", () => {
//   const handleRowClick = jest.fn();
//   const sut = shallow(
//     <Autocomplete
//       data={[{ a: 1, b: 2 }]}
//       desc={false}
//       onRowClick={handleRowClick}
//     />
//   );

//   sut
//     .find("tbody tr")
//     .at(0)
//     .simulate("click");

//   expect(handleRowClick).toBeCalled();
// });

// it("should set row props", () => {
//   const sut = shallow(
//     <TablePlain
//       data={[{ a: 1, b: 2 }]}
//       desc={false}
//       rowProps={() => ({
//         style: { background: "yellow" },
//         className: "catchMeIfYouCan"
//       })}
//     />
//   );

//   const row = sut.find(".catchMeIfYouCan").at(0);

//   expect(row.props()).toHaveProperty("style.background");
// });

// it("should call renderRoot if provided", () => {
//   const sut = shallow(
//     <TablePlain
//       data={[]}
//       desc={false}
//       renderRoot={children => <div className="table">{children}</div>}
//     />
//   );

//   expect(sut.find(".table").length).toBe(1);
// });

// describe("subComponent", () => {
//   it("should not call onRowClick when clicked", () => {
//     const handleRowClick = jest.fn();
//     const sut = mount(
//       <TablePlain
//         data={[{ a: 1, b: 2 }]}
//         desc={false}
//         subComponent={() => <h1>Dummy</h1>}
//         onRowClick={handleRowClick}
//       />
//     );

//     sut
//       .find(".expander") // Click on Expander
//       .first()
//       .simulate("click");

//     expect(handleRowClick).not.toBeCalled();
//   });
// });
