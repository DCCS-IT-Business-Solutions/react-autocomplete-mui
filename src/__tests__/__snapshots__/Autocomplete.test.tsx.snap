// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {
          "class": "MuiFormControl-root MuiTextField-root",
          "style": "min-width:500px",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "aria-invalid": "false",
                  "class": "MuiInputBase-input MuiInput-input",
                  "type": "text",
                  "value": "",
                },
                "children": Array [],
                "name": "input",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "aria-invalid": undefined,
                  "class": undefined,
                  "type": undefined,
                  "value": undefined,
                },
                "x-attribsPrefix": Object {
                  "aria-invalid": undefined,
                  "class": undefined,
                  "type": undefined,
                  "value": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
